fdbserver/ptxn/test/RealStorageServer.actor.cpp

```diff
@@ -0,0 +1,146 @@
+/*
+ * RealStorageServer.actor.cpp
+ *
+ * This source file is part of the FoundationDB open source project
+ *
+ * Copyright 2013-2021 Apple Inc. and the FoundationDB project authors
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#include "fdbserver/MockPeekCursor.h"
+#include "fdbserver/MockLogSystem.h"
+#include "fdbclient/StorageServerInterface.h"
+#include "fdbserver/IKeyValueStore.h"
+#include "fdbserver/WorkerInterface.actor.h"
+#include "fdbserver/RecoveryState.h"
+#include "fdbserver/ptxn/test/Driver.h"
+
+#include "flow/UnitTest.h"
+
+#include "flow/actorcompiler.h" // has to be last include
+
+namespace ptxn {
+
+ACTOR Future<Void> driverRunStorageServer(StorageServerInterface* ssi,
+                                          std::vector<Future<Void>>* actors,
+                                          Tag tag,
+                                          Reference<MockPeekCursor> mockPeekCursor) {
+
+	KeyValueStoreType storeType = KeyValueStoreType::SSD_BTREE_V2;
+	state std::string folder = ".";
+	std::string fileName = joinPath(folder, "storage-" + ssi->id().toString() + "." + storeType.toString());
+	std::cout << "new Storage Server file name: " << fileName << std::endl;
+	deleteFile(fileName);
+	state IKeyValueStore* data = openKVStore(storeType, fileName, ssi->id(), 0);
+
+	state ReplyPromise<InitializeStorageReply> storageReady;
+
+	// Initialize dbInfo with mocked cursor and mocked loy system.
+
+	// TODO: Use Reference API?
+	// TODO: delete mockLogSystem eventually.
+	state MockLogSystem* mockLogSystem = new MockLogSystem();
+	mockLogSystem->cursor = mockPeekCursor.castTo<ILogSystem::IPeekCursor>();
```

Comment:
It looks like mockLogSystem emulates higher layers so peek can get transactions to run in the SS.

Another way might be to [reuse code here](https://github.com/apple/foundationdb/blob/0ec1c9deb2b7943004d199f619369ce66b725fc8/fdbserver/ptxn/test/RealTLog.actor.cpp#L243) to create an actual stand-alone LogSystem and then call the TLogCommit interface, then after that start the SS to do the peeks. This is how "real" TLog was tested and avoids having to modify existing files or create new ones. So is less intrusive (of course in a branch thats not a big deal). I think it would be nice if the 'real' tests were more similar and maybe work together.

Reply:
>IIUC, the point of having component level tests is to avoid depending on the implementation of the other components, how can reusing ILogSystem::fromServerDBInfo achieve this?

my own implementation choice was to try to find where tLog is spending the most time by running actual code as much as possible. i.e. mock stuff outside the component but avoiding mocking stuff inside the component (including whatever the component calls) to get performance numbers close to reality.